torch/_export/passes/add_runtime_assertions_for_constraints_pass.py

```diff
@@ -0,0 +1,191 @@
+from collections import defaultdict, namedtuple
+from typing import Dict, List, Tuple
+
+import math
+import operator
+import sympy
+
+import torch.utils._pytree as pytree
+import torch
+import torch.fx
+from torch._export.pass_base import ExportPassBase, ProxyValue
+from torch._export.graph_module import get_export_meta
+from torch._export.pass_infra.node_metadata import NodeMetadata
+
+
+__all__ = ["AddRuntimeAssertionsForConstraintsPass"]
+
+
+ConstraintSpec = namedtuple("ConstraintSpec", ["constraint_dim", "min_val", "max_val"])
+
+
+class AddRuntimeAssertionsForConstraintsPass(ExportPassBase):
+    def __init__(self) -> None:
+        super().__init__()
+        self.current_gm = None
+
+    def _process_constraints(self, constraints) -> Dict[int, List[ConstraintSpec]]:
+        constraints_id_to_constraint: Dict[int, List[ConstraintSpec]] = defaultdict(
+            list
+        )
+        if constraints is None:
+            return constraints_id_to_constraint
+
+        # Convert simple sympy Integers into concrete int to
+        # insert into graph
+        def _convert_to_int(val):
+            if val == sympy.oo:
+                return math.inf
+            if isinstance(val, sympy.Integer):
+                return int(val)
+            raise RuntimeError(
+                "Export constraints cannot be non-integer expressions"
+            )
+
+        constraint_id_to_dim: Dict[int, Dict[int, List[Tuple[int, int]]]] = defaultdict(
+            lambda: defaultdict(list)
+        )
+        for constraint in constraints:
+            min_int_val = _convert_to_int(constraint["min"])
+            max_int_val = _convert_to_int(constraint["max"])
+            constraint_id_to_dim[constraint["t_id"]][constraint["dim"]].append(
+                (min_int_val, max_int_val)
+            )
+
+        # Merge the constraints into a single list of constraints
+        for t_id in constraint_id_to_dim:
+            for dim, constraints in constraint_id_to_dim[t_id].items():
+                min_vals = [x[0] for x in constraints]
+                max_vals = [x[1] for x in constraints]
+                min_val = sorted(min_vals, reverse=True)[0]
+                max_val = sorted(max_vals, reverse=False)[0]
+
+                assert min_val <= max_val
+
+                constraints_id_to_constraint[t_id].append(
+                    ConstraintSpec(constraint_dim=dim, min_val=min_val, max_val=max_val)
+                )
```

Comment:
All the processing here is to reorganize the constraints by input tensor. I'm not sure why the `Constraint` class is not defined as a ref to a single tensor with a *list* of dims and ranges in the very first place.

Reply:
I'm totally fine with it.